### ISSUE TITLE
AppArmor: Samba profiles always get switched to complain mode

### ISSUE THREAD

**Issue by cboltz** [Nobara-Project/rpm-sources#116]
In the AppArmor package, there's a patch that always sets the samba profiles to complain mode - without giving the user an option to override this:

[baseos/apparmor/set_samba_profiles_to_complain_mode_in_ExecStartPre.patch](https://github.com/Nobara-Project/rpm-sources/blob/main/baseos/apparmor/set_samba_profiles_to_complain_mode_in_ExecStartPre.patch)

This looks like a bad idea to me. If there isn't a very good reason for this, I'd propose to drop this patch.

(Oh, and the %changelog looks terribly outdated ;-) and sadly doesn't explain why this patch exists.)

BTW: I'm one of the upstream AppArmor developers, and also do the openSUSE packaging. If you have questions or need help, feel free to ask ;-)

**Comment by GloriousEggroll** [comment]
> In the AppArmor package, there's a patch that always sets the samba profiles to complain mode - without giving the user an option to override this:
> 
> [baseos/apparmor/set_samba_profiles_to_complain_mode_in_ExecStartPre.patch](https://github.com/Nobara-Project/rpm-sources/blob/main/baseos/apparmor/set_samba_profiles_to_complain_mode_in_ExecStartPre.patch)
> 
> This looks like a bad idea to me. If there isn't a very good reason for this, I'd propose to drop this patch.
> 
> (Oh, and the %changelog looks terribly outdated ;-) and sadly doesn't explain why this patch exists.)
> 
> BTW: I'm one of the upstream AppArmor developers, and also do the openSUSE packaging. If you have questions or need help, feel free to ask ;-)

sure, our apparmor implementation is bare minimum, basically we just wanted -something- in place so that we could get away from selinux.

As for the samba change -- here's the entire discord thread as to why that change was made:


```
Osaris — 03/12/2023 2:39 PM
its definitely apparomor

Osaris — 03/12/2023 2:41 PM
type=USER_ACCT msg=audit(1678653359.973:869): pid=15724 uid=1000 auid=1000 ses=4 subj=unconfined msg='op=PAM:accounting grantors=pam_unix,pam_localuser acct="osaris" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/2 res=success'UID="osaris" AUID="osaris"
type=USER_CMD msg=audit(1678653359.973:870): pid=15724 uid=1000 auid=1000 ses=4 subj=unconfined msg='cwd="/home/osaris" cmd=61612D636F6D706C61696E2073616D62612D646365727063642073616D62612D626771642073616D62612D727063642073616D62612D727063642D636C61737369632073616D62612D727063642D73706F6F6C7373 exe="/usr/bin/sudo" terminal=pts/2 res=success'UID="osaris" AUID="osaris"
type=CRED_REFR msg=audit(1678653359.973:871): pid=15724 uid=1000 auid=1000 ses=4 subj=unconfined msg='op=PAM:setcred grantors=pam_env,pam_localuser,pam_unix acct="root" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/2 res=success'UID="osaris" AUID="osaris" (edited)
[2:42 PM]
type=USER_START msg=audit(1678653359.974:872): pid=15724 uid=1000 auid=1000 ses=4 subj=unconfined msg='op=PAM:session_open grantors=pam_keyinit,pam_limits,pam_keyinit,pam_limits,pam_systemd,pam_unix acct="root" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/2 res=success'UID="osaris" AUID="osaris"
type=AVC msg=audit(1678653360.081:873): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="samba-dcerpcd" pid=15727 comm="apparmor_parser"
type=AVC msg=audit(1678653360.096:874): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="samba-bgqd" pid=15729 comm="apparmor_parser"
type=AVC msg=audit(1678653360.110:875): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="samba-rpcd" pid=15731 comm="apparmor_parser"
type=AVC msg=audit(1678653360.124:876): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="samba-rpcd-classic" pid=15733 comm="apparmor_parser"
type=AVC msg=audit(1678653360.140:877): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="samba-rpcd-spoolss" pid=15735 comm="apparmor_parser"
type=USER_END msg=audit(1678653360.150:878): pid=15724 uid=1000 auid=1000 ses=4 subj=unconfined msg='op=PAM:session_close grantors=pam_keyinit,pam_limits,pam_keyinit,pam_limits,pam_systemd,pam_unix acct="root" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/2 res=success'UID="osaris" AUID="osaris"
type=CRED_DISP msg=audit(1678653360.151:879): pid=15724 uid=1000 auid=1000 ses=4 subj=unconfined msg='op=PAM:setcred grantors=pam_env,pam_localuser,pam_unix acct="root" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/2 res=success'UID="osaris" AUID="osaris" (edited)
[2:42 PM]
fixed with suggestions from https://forum.manjaro.org/t/root-tip-how-to-basic-samba-setup-and-troubleshooting/100420
[2:42 PM]
sudo aa-complain /etc/apparmor.d/usr.sbin.smbd (edited)
[2:42 PM]
sudo aa-complain samba-dcerpcd samba-bgqd samba-rpcd samba-rpcd-classic samba-rpcd-spoolss (edited)
[2:43 PM]
i can now create the shares
[2:43 PM]
to recreate - install the Samba Filesharing Plugin  https://apps.kde.org/kdenetwork_filesharing/ via kde discover
sudo dnf install samba
sudo systemctl enable smb nmb --now
sudo firewall-cmd --permanent --add-service=samba
sudo firewall-cmd --reload

sudo mkdir /var/lib/samba/usershares
sudo groupadd -r sambashare
sudo chown root:sambashare /var/lib/samba/usershares
chmod 1770 /var/lib/samba/usershares

In /etc/samba/smb.conf add:

[global]
  usershare path = /var/lib/samba/usershares
  usershare max shares = 100
  usershare allow guests = yes
  usershare owner only = yes

Run:

sudo gpasswd sambashare -a your_username
sudo systemctl restart smb nmb

reboot. (for some reason fedora does not pickup the group additions to your user account and not sure why but logout does not work and you have to reboot thus you wont have permissions to /var/lib/samba/usershares until you reboot) (edited)
[2:44 PM]
right click on whatever you want to share, try and share it then you get this error
[2:46 PM]
An error ocurred while trying to share diectory. The share has not been created.
Samba internals report:
net usershare add: cannot convert name "Everyone" to a SID. The transport connection is now disconnected..

Osaris — 03/12/2023 2:55 PM
quick way to test for the error after setting up the smb.conf  is with  net usershare add --long share ~/testshare "this is a comment" "Everyone:R" guest_ok=n
[2:57 PM]
as i am a noob with apparmor its possible that new apparmor profiles are needed for samba. This might or might not help, was too lazy to try and figure out what they are saying "https://gitlab.com/apparmor/apparmor/-/merge_requests/871"

Osaris — 03/12/2023 3:12 PM
the samba stuff was in enforcing mode in apparmor  which can be viewed with sudo apparmor_status i installed the kde version of nobara and the installation is about 3 days old and i have not touched apparmor myself until now. (edited)
March 15, 2023

Osaris — 03/15/2023 1:22 PM
@GloriousEggroll
[1:26 PM]
this will benefit users that setup smb shares. would prefer an even easier way todo it to accommodate new users. OpenSuse Leap has done this quite good if it can be of any use as a reference point to check what they did.

GloriousEggroll — 03/15/2023 1:40 PM
@Osaris thanks i can modify the apparmor profile for samba and make a nobara wiki page for the rest of the config
[1:40 PM]
will do so when i get some time

Osaris — 03/15/2023 1:41 PM
no problem at all. thank you for all your hard work!

GloriousEggroll — 03/15/2023 7:04 PM
@Osaris those profiles were merged into upstream apparmor 11 months ago
[7:04 PM]
our version is 3.1.2 which already includes them

GloriousEggroll — 03/15/2023 9:15 PM
i just went ahead and added the aa-complain lines to the post section of the installer
March 16, 2023

Osaris — 03/16/2023 11:21 AM
awesomeness!
```

**Comment by cboltz** [comment]
Thanks for the additional details!

What you pasted unfortunately doesn't include the DENIED log lines (or ALLOWED in complain mode), but it explains the problems you had nevertheless.

Basically there are two things you need:

a) read and write access to `/var/lib/samba/usershares/`

That's included in the smbd profile in AppArmor 3.0.1 and newer

b) permissions for the actual shares so that users can access them

This was a repeated topic for openSUSE bugreports - until I wrote a little script that updates the AppArmor profile for smbd based on smb.conf. You can find the `update-apparmor-samba-profile` script in the openSUSE and Debian packages - or you just take it from here: [update-apparmor-samba-profile script](https://github.com/user-attachments/files/17780426/update-apparmor-samba-profile.txt)
This script should be run whenever Samba gets restarted, ideally in smb.service:
`ExecStartPre=/usr/share/samba/update-apparmor-samba-profile`

I'd recommend to include this script in smb.service (or a drop-in if you want to ship it via the AppArmor package) instead of switching the Samba profiles to complain mode.

Note: I have to admit that I never tested with usershares, so maybe the script needs an addition to handle them.
(The default profile already allows sharing /home/*/, so the most likely usecases for usershares, "share something inside my home dir", should already be covered.)

If you try this and run into denials, just show me the log lines so that i can help to fix the profiles.

**Comment by GloriousEggroll** [comment]
fixed in latest sources for N41
